src/display/xfa_layer.js

```diff
@@ -0,0 +1,89 @@
+/* Copyright 2021 Mozilla Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+class XfaLayer {
+  static setAttributes(html, attrs) {
+    for (const [key, value] of Object.entries(attrs)) {
```

Comment:
Just to confirm, only the predefined attributes from the various toHTML can be set here right? I just want to confirm we don't need need to remove any potentially malicious attributes from XFA (e.g. onclick).

Reply:
Yeah, normally everything is under control:
  - only nodes and attributes defined in spec are in the final form DOM;
  - the various `toHTML` functions will generate some HTML based on those attributes and nodes.
So it's impossible to have some injections.
XFA can contain some xhtml, by spec nodes and attributes are limited too. And for xhtml nodes style is parsed to filter out any forbidden property:
https://github.com/mozilla/pdf.js/blob/master/src/core/xfa/xhtml.js#L63
So it should be safe.